sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -283,3 +303,166 @@ def get_lpmode_support(self):
+        # The software reset bit is self-clear, write 1 to initiate the RESET
+        # and it will be cleared to 0 by the hardware when out of RESET.
+        ret = self.xcvr_eeprom.write(consts.MODULE_LEVEL_CTRL_RESET_FIELD, 1)
+        if ret:
+            for i in range(CMIS_RESET_TIMEOUT * 2):
+                time.sleep(0.5)
+                val = self.xcvr_eeprom.read(consts.MODULE_LEVEL_CTRL_RESET_FIELD)
+                # Device is in REST, the I2C read could fail, and None could be returned
+                if (val is None) or val:
+                    ret = False
+                    continue
+                # Device is out of REST
+                ret = True
+                break
+        return ret
+
+    def get_error_description(self):
+        val = self.xcvr_eeprom.read(consts.DATAPATH_STATE_FIELD)
+        if val not in [0, CMIS_DATAPATH_STATE_ACTIVATED]:
+            return SFP_ERROR_DESCRIPTION_CMIS_DATAPATH_ERROR
+
+        val = self.xcvr_eeprom.read(consts.CONFIG_ERROR_FIELD)
+        if val not in [0, CMIS_CONFIG_STATE_ACCEPTED]:
+            return SFP_ERROR_DESCRIPTION_CMIS_CONFIG_ERROR
+
+        state = self.xcvr_eeprom.read(consts.MODULE_STATE_FIELD)
+        if state not in [CMIS_MODULE_STATE_NONE, CMIS_MODULE_STATE_READY]:
+            return SFP_ERROR_DESCRIPTION_CMIS_NOT_READY
+
+        return None
+
+    def get_host_speed(self, ifname):
+        # see host_electrical_interface of sff8024.py
+        speed = 0
+        if '400G' in ifname:
+            speed = 400000
+        elif '200G' in ifname:
+            speed = 200000
+        elif '100G' in ifname or 'CAUI-4' in ifname:
+            speed = 100000
+        elif '50G' in ifname or 'LAUI-2' in ifname:
+            speed = 50000
+        elif '40G' in ifname or 'XLAUI' in ifname or 'XLPPI' in ifname:
+            speed = 40000
+        elif '25G' in ifname:
+            speed = 25000
+        elif '10G' in ifname or 'SFI' in ifname or 'XFI' in ifname:
+            speed = 10000
+        elif '1000BASE' in ifname:
+            speed = 1000
+        return speed
+
+    def get_cmis_application_selected(self):
+        ap_code = 0
+        if not self.is_flat_memory():
+            field = "{}_{}".format(consts.STAGED_CTRL0_APSEL_FIELD, 1)
+            ap_code = self.xcvr_eeprom.read(field) >> 4
+
+        return (ap_code & 0xf)
+
+    def get_cmis_application_matched(self, host_speed, host_lanes):
+        if host_speed == 0 or host_lanes == 0:
+            return 0
+
+        ap_code = 0
+        appl_dict = ast.literal_eval(self.xcvr_eeprom.read(consts.APPLS_ADVT_FIELD))
+        for c in appl_dict.keys():
+            d = appl_dict[c]
+            if d.get('host_lane_count') != host_lanes:
+                continue
+            if self.get_host_speed(d.get('host_electrical_interface_id')) != host_speed:
+                continue
+            ap_code = c
+
+        return (ap_code & 0xf)
+
+    def set_cmis_application(self, host_speed, host_lanes):
```

Comment:
how are different speeds on different lane blocks configured?

Reply:
This is not a supported port mode in the Broadocm, hence we can't validate it, and having a different port mode (lane count, speed) on a QSFPDD requires SERDES support on the MAC, and it is outside the scope of this PR.

Reply:
When you say "This is not a supported port mode in the Broadocm", what level support are you talking about? Is this Broadcom SONiC? I believe this should be supported at the SDK/ASIC layer.

Reply:
Sorry, I was referring to the CMIS transceivers available to me, none of them are able to support different application codes on lanes of the single optic, and yes, this is also not a supported port mode on Broadcom SONIC switches, hence I can't do the verification.
Since we support only default application as of now, we'll deal with it later when the synchronization mechanism between syncd and xcvrd is in place and ready